compiler/rustc_mir/src/transform/instrument_coverage.rs

```diff
@@ -54,69 +50,131 @@ pub(crate) fn provide(providers: &mut Providers) {
 /// are still included in the total `num_counters` or `num_expressions`.) Simply counting the
 /// calls may not work; but computing the number of counters or expressions by adding `1` to the
 /// highest ID (for a given instrumented function) is valid.
+///
+/// This visitor runs twice, first with `add_missing_operands` set to `false`, to find the maximum
+/// counter ID and maximum expression ID based on their enum variant `id` fields; then, as a
+/// safeguard, with `add_missing_operands` set to `true`, to find any other counter or expression
+/// IDs referenced by expression operands, if not already seen.
```

Comment:
What keeps us from doing this in one pass? If we're falling back to the MAX_COUNTER_GUARD anyway it seems like we could always use that?

Reply:
I need to revisit this when I get back. My first thought was, no because I'm only forcing the counters to be less than the max, but expressions could be larger than u32:MAX minus the max counters. But if that happens then counters really can't reach the max counters constant so using that probably won't work anyway. That means I would have to actually pick a max value and stick to it for both counters and expressions, which I don't really like because I don't know what the right split should be, and it could change over time. If I can't come up with a way to dynamically figure this out like I used to, then I think I may just have to do that, in which case yes maybe I can make it just one pass here.